kiwi/lib/search.py

```diff
@@ -0,0 +1,655 @@
+    if base_config_dict['trainer']['main_metric'] is None:
+        raise ValueError(
+            'The metric should be explicitly set in `trainer.main_metric` '
+            'in the training config (`base_config`).'
+        )
+    # The main metric(s) should be inside a list
+    if not isinstance(base_config_dict['trainer']['main_metric'], list):
+        base_config_dict['trainer']['main_metric'] = [
+            base_config_dict['trainer']['main_metric']
+        ]
+
+    # Use the early stopping logic of the Kiwi trainer
+    base_config_dict['trainer']['checkpoint'][
+        'early_stop_patience'
+    ] = config.options.patience
+    base_config_dict['trainer']['checkpoint'][
+        'validation_steps'
+    ] = config.options.validation_steps
+
+    # Load or initialize a study
+    if config.load_study:
+        logger.info(f'Loading study to resume from: {config.load_study}')
+        study = joblib.load(config.load_study)
+    else:
+        if config.options.search_method == 'random':
+            logger.info('Exploring parameters with random sampler')
+            sampler = optuna.samplers.RandomSampler(seed=config.seed)
+        else:
+            multivariate = (config.options.search_method == 'multivariate_tpe',)
+            logger.info(
+                'Exploring parameters with '
+                f'{"multivariate " if multivariate else ""}TPE sampler'
+            )
+            sampler = optuna.samplers.TPESampler(
+                seed=config.seed, multivariate=multivariate,
+            )
+        logger.info('Initializing study...')
+        pruner = optuna.pruners.MedianPruner()
+        study = optuna.create_study(
+            study_name=config.search_name,
+            direction='maximize',
+            pruner=pruner,
+            sampler=sampler,
+        )
+
+    # Optimize the study
+    optimize_kwargs = {}
+    if use_mlflow:
+        optimize_kwargs['callbacks'] = [optuna.integration.MLflowCallback()]
+    try:
+        logger.info('Optimizing study...')
+        objective = Objective(config=config, base_config_dict=base_config_dict)
+        study.optimize(
+            objective, n_trials=config.num_trials, **optimize_kwargs,
+        )
+    except KeyboardInterrupt:
+        logger.info('Early stopping search caused by ctrl-C')
+    except Exception as e:
+        logger.error(
+            f'Error occured during search: {e}; '
+            f'current best params are: {study.best_params}'
+        )
+
+    logger.info(f"Saving study to: {output_dir / 'study.pkl'}")
+    joblib.dump(study, output_dir / 'study.pkl')
+
+    try:
+        logger.info('Number of finished trials: {}'.format(len(study.trials)))
+        logger.info('Best trial:')
+        trial = study.best_trial
+        logger.info('  Value: {}'.format(trial.value))
+        logger.info('  Params: ')
+        for key, value in trial.params.items():
+            logger.info('    {}: {}'.format(key, value))
+    except Exception as e:
```

Comment:
Same as above. Why catch a generic exception and throw a generic error message? You either catch a real exception or don't catch them at all. This occasion is more critical because you are no longer protecting a critical functionality (saving the study)